### PR TITLE
exSpy Split

### PR DESCRIPTION
**Pull request by CSSFrancis**
### Description of the change
This is my (best) attempt at splitting the EDS and EELS. 

### Progress of the PR
- [x] Move Signals to exspy
- [x] Move Models to exspy
- [x] Move Components to exspy
- [x] Move EdgesRange to exspy (Wait till Marker Collection is Merged?)
- [x] Move Utils to espy
- [x] Move EELS/EDS Drawing tests to Spectrospy (Wait till Marker Collection is Merged?)
- [x] Move EDS/EELS Tests
- [x] Move model convolution to exSpy
- [x] Skip all Tests in hyperspy if exspy isn't installed and required.
- [x] Get Tests back to passing 
- [x] Move Documentation
- [x] Rename to exSpy

Current State:

Any PRS are welcome here.  Most of the splitting/ testing is done but there is quite a bit in the hyperspy tests that are still dependent on spectrospy.  I don't think that is a huge deal as we can always split, require spectrospy as a testing requirement and then clean up those tests later.

### CODE REVIEW

**Comment by CSSFrancis** [comment]
This appears to be failing due to the GUI toolkit.  I'll need to look into this a little bit more...

**Comment by CSSFrancis** [comment]
@ericpre for some reason the spectrospy package isn't getting loaded properly. I think it is because of the GUI components but I have the annoying problem of it not failing on my computer... 

Do you have any ideas about how to fix this?

**Comment by CSSFrancis** [comment]
My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.

**Comment by francisco-dlp** [comment]
Great job!

Yesterday @ericpre and I discussed offline the name of the splitted package. We thought that it would be good that the name incorporates the idea that it is an EM package, basically devoted to EELS and EDX. So, @ericpre suggested adding an `e` (electron, EELS) (`EspectroSpy`), and I suggested replacing the first `s` with `x` (EDX) `EXpectroSpy`. What do you think?

**Comment by CSSFrancis** [comment]
@francisco-dlp I don't have terribly strong opinions on any name really. I'm not really a huge fan of spectrospy cause it is very close to spectroscopy, I just needed a place holder for now.

Something shorter might be nice even if it's not the most accurate.

- Exspy
- Espy
- Spexspy

Expectrospy sounds like a spell from Harry Potter :) 

My vote would be for Expectrospy or just exspy or espy.

I like the exspy maybe the best because it also gives the idea that it used to be part of hyperspy.

**Comment by ericpre** [comment]
> Expectrospy sounds like a spell from Harry Potter :)

Agree, and it is rather on the long side!

`EXSpy` sounds good to me and it is easy to read/pronounce (EX-Spy)?
An alternative may be `EelXSpy` (pronounciation: Eel-X-Spy; il-ɛks-spaɪ) or `EleXSpy` (EleX-Spy; ɛlɛks-spaɪ). I find the latter is slightly easier to pronounce.

> I like the exspy maybe the best because it also gives the idea that it used to be part of hyperspy.

Yes, but this can also be misleading/distracting.

**Comment by jlaehne** [comment]
Thumbs up for *EleXSpy*

For me the *Spy* already contains spectroscopy and python, so anything with spectro would be a redundancy.

**Comment by francisco-dlp** [comment]
I like the minimalist ex*Spy*: it contains all the relevant information (and maybe even too much) in the minimum number of characters.

**Comment by ericpre** [comment]
We are close to be able to come back to splitting but it seems that we still two contenders for the package name: `exSpy` or `EleXSpy`, should we put it to a vote? 

| name             |  ballot paper  |
|---------------|-----------------|
| `exSpy`          |  :rocket:          |
| `eleXSpy`       |  :tada:            |

**Comment by CSSFrancis** [comment]
@ericpre I think I still need to adjust the minimum test to run with the gui packages.

Other than that I think most of the PR is getting closer to being done. There is a bit more cleaning up to do especially with the documentation.  I think a lot of it might stay for now as I think they are still good examples and the hyperspy documentation isn't really complete without them.

We still need to change the name of the package as well but that should be rather minimal. 

I don't if I'll be able to get back to this until Thursday morning so if anyone else wants to take this up feel free otherwise I will try to finish this then.

**Comment by CSSFrancis** [comment]
@ericpre do you want to look at this when you get the chance?  

I think that it is good to go.  At least most of the functionality has been split.  There are a couple of things that I don't know if I should move over such as the hard coded elements `hyperspy.misc.elements` and `hyperspy.misc.physics_tools` 

Part of me thinks that these parts of the code might be better in the eels package but I'm not sure.  At this point hyperspy 
is just Visualization, ML and some parallel computing. I'm pretty happy with that current state even if I think that the ML section needs a little triage ;)

**Comment by CSSFrancis** [comment]
Just a fun little comment but after splitting the io the eels/eds and holography:
```
+----------+-----------------+-----------------+----------------------+
| Package  |  Lines of Code  |  Lines of Tests |   Lines of Comments  |
+----------+-----------------+-----------------+----------------------+
| Hyperspy |       33,889    |      22,810     |       22,805         |
+----------+-----------------+-----------------+----------------------+
| exspy    |        5,721    |       2,771     |        4,282         |
+----------+-----------------+-----------------+----------------------+
| rsciio   |       17,453    |      15,971     |        6,815         |
+----------+-----------------+-----------------+----------------------+
| pyxem    |       10,897    |      13,356     |       11,788         |
+----------+-----------------+-----------------+----------------------+
| holospy  |          603    |         371     |          519         |
+----------+-----------------+-----------------+----------------------+
```

This means that hyperspy is about half the size from when it started which should make changes to the code base much more manageable. These numbers are definitely effected by `black` but I still think this is a great change and really good for the health of the community. 

I realize that lines of code are a bad way to measure things but, I think that at this point there is a pretty good argument that hyperspy is in fact __not bloated__ as has been suggested to me in the past. ;) 

The release of hyperspy 2.0.0 should also remove a fair bit of code from pyxem meaning that it will likely shrink with time 😄

**Comment by ericpre** [comment]
`hyperspy.misc.elements` needs to go in exspy package, it shouldn't stay in hyperspy. `hyperspy.misc.physics_tools` is not used anywhere and should be removed in a separate PR, it is trivial code that doesn't have its place in hyperspy (or exspy).

**Comment by CSSFrancis** [comment]
> `hyperspy.misc.elements` needs to go in exspy package, it shouldn't stay in hyperspy. `hyperspy.misc.physics_tools` is not used anywhere and should be removed in a separate PR, it is trivial code that doesn't have its place in hyperspy (or exspy).

Okay I'll pull out the elements and leave the physics tools for a separate PR.  I need to clean up my commit history a bit and fix the failing tests.  I was a little too confident before pushing some of the changes 😅

**Comment by ericpre** [comment]
Where does the `exspy.gui`  module come from? It is something that comes from the git history when splitting?

File: exspy/exspy/misc/eds/eds.py [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 # GNU General Public License for more details.
 #
 # You should have received a copy of the GNU General Public License
-# along with HyperSpy. If not, see <https://www.gnu.org/licenses/#GPL>.
+# along with exspy. If not, see <https://www.gnu.org/licenses/#GPL>.
 
-from hyperspy.misc.eds.utils import (
+
+from exspy.exspy.misc.eds.utils import (
```
Comment (ericpre) [inline exspy/exspy/misc/eds/eds.py]:
```suggestion
from exspy.misc.eds.utils import (
```
?

**Review by ericpre — COMMENTED** [review]
Should the `EDS_SEM_Spectrum` and `EDS_TEM_Spectrum` dataset be moved too? It is a bit of a pain because it is used in example, but it does need to go, I think!

**Comment by CSSFrancis** [comment]
> Where does the `exspy.gui` module come from? It is something that comes from the git history when splitting?

This comes from the `hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` packages.  Maybe this causes problems related  running with no_gui but I'm not sure.  Otherwise a separate `gui`  package for exspy with only a little code seems silly. 

Maybe it's just complicated as we are splitting things but I was getting frustrated with juggling 6 different repos and getting the tests to pass. 

> My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.

This was my idea for how to get things to work?

**Comment by CSSFrancis** [comment]
> Should the `EDS_SEM_Spectrum` and `EDS_TEM_Spectrum` dataset be moved too? It is a bit of a pain because it is used in example, but it does need to go, I think!

Yes they should be moved...  We need to think of the best way to not have our test coverage drop.  We can move them and still depend on them for testing but __ideally__ the tests wouldn't depend on that.

**Comment by CSSFrancis** [comment]
@ericpre Okay sorry for the confusion :) Everything __should be__ more clean now (and hopefully the tests actually pass)

**Comment by ericpre** [comment]
> Yes they should be moved... We need to think of the best way to not have our test coverage drop. We can move them and still depend on them for testing but ideally the tests wouldn't depend on that.

The test needs to be updated to use generic signal. It doesn't seem to be used in many places, so not too much of a pain!

File: exspy/doc/intro.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
+
+What is Spectrospy
```
Comment (ericpre) [inline exspy/doc/intro.rst]: Do you want to rename `Spectrospy` to `exSpy` now or leave to when the repository is setup?

**Comment by ericpre** [comment]
> > Where does the `exspy.gui` module come from? It is something that comes from the git history when splitting?
> 
> This comes from the `hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` packages. Maybe this causes problems related running with no_gui but I'm not sure. Otherwise a separate `gui` package for exspy with only a little code seems silly.
> 
> Maybe it's just complicated as we are splitting things but I was getting frustrated with juggling 6 different repos and getting the tests to pass.
> 
> > My current plan is that we should be able to `git-filter-repo` the commits from `hyperspy`,` hyperspy_gui_traitsui` and `hyperspy_gui_ipywidgets` so that all of the EELS EDS stuff is in one place.
> 
> This was my idea for how to get things to work?

I was expecting to leave the gui code in their respective packages and not touch it? But yes, both approach are possible! I have a slight preference to keep it as it is because it means that the `exSpy` doesn't have to deal with gui dependencies at all and for maintenance (ci, running test suite) and installation, it is easier.

File: hyperspy/tests/signals/test_assign_subclass.py [hunk 3]
```diff
@@ -141,27 +136,31 @@ def setup_method(self, method):
         self.s = hs.signals.Signal1D([0, 1])
 
     def test_lazy_to_eels_and_back(self):
+        exspy = pytest.importorskip("exspy")
```
Comment (ericpre) [inline hyperspy/tests/signals/test_assign_subclass.py]: I think these tests are redundant, in hyperspy, we only need one test to check that setting a `signal_type` change the object class but that's all.

Testing assignment to signal specify class should go in `exSpy`

**Review by ericpre — COMMENTED** [review]
We need to move the preferences too:

https://github.com/hyperspy/hyperspy/blob/70ca0762a332a7239a4c1eb75001062ab6c7495d/hyperspy/defaults_parser.py#L119-L123

https://github.com/hyperspy/hyperspy/blob/70ca0762a332a7239a4c1eb75001062ab6c7495d/hyperspy/defaults_parser.py#L180-L199

**Comment by CSSFrancis** [comment]
> I was expecting to leave the gui code in their respective packages and not touch it? But yes, both approach are possible! I have a slight preference to keep it as it is because it means that the `exSpy` doesn't have to deal with gui dependencies at all and for maintenance (ci, running test suite) and installation, it is easier.

We can't leave it in their respective packages or else the gui for just hyperspy doesn't work without exspy installed? Or do I have that wrong and the extensions are happy when  there are functions without toolkits? 

We can always have the gui stuff in exspy and then have it be an optional install? Correct? or does that make all of the tests harder to run.  I'll admit the gui toolkit stuff has been kind of a pain to split so I'm open to all suggestions. 

> The test needs to be updated to use generic signal. It doesn't seem to be used in many places, so not too much of a pain!

Can you do that possibly?  I don't have a great understanding of the peak fitting stuff in `hyperspy` so writing new tests takes me a bit longer than it should :)

**Comment by ericpre** [comment]
> We can't leave it in their respective packages or else the gui for just hyperspy doesn't work without exspy installed? Or do I have that wrong and the extensions are happy when there are functions without toolkits?

I haven't check but I would expect that it would work! Using `hyperspy_gui_*` as it is without having `expy` installed should work through the registry: each library declare their class/function in the `ymal` file, these get registered and these are when necessary.

> Can you do that possibly? I don't have a great understanding of the peak fitting stuff in hyperspy so writing new tests takes me a bit longer than it should :)

Yes, I can do this.

**Comment by CSSFrancis** [comment]
> I haven't check but I would expect that it would work! Using `hyperspy_gui_*` as it is without having `expy` installed should work through the registry: each library declare their class/function in the `ymal` file, these get registered and these are when necessary.

@ericpre Hmm okay, I was kind of assuming that was the case when I started but for some reason it wasn't working.  That's probably just because something got moved and the whole `.ymal`  extensions got messed up.  Getting everything to talk nicely with lots of different version etc has been not the easiest thing which is why I changed things. 

I might not get around to giving this another shot until the weekend. If you want to take a crack at it feel free.

**Comment by ericpre** [comment]
Okay, I will give a go.

File: exspy/exspy/components/eels_arctan.py [hunk 4]
```diff
@@ -1,20 +1,21 @@
 # -*- coding: utf-8 -*-
-# Copyright 2007-2023 The HyperSpy developers
+# Copyright 2007-2023 The exspy developers
```
Comment (jlaehne) [inline exspy/exspy/components/eels_arctan.py]: In some file headers, `exSpy` is capitalized correctly, but in most it is still all lower case.

File: exspy/exspy/release_info.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+name = "exspy"
+version = "0.0.1"
+author = "exspy Developers"
```
Comment (jlaehne) [inline exspy/exspy/release_info.py]:
```suggestion
author = "exSpy Developers"
```

File: exspy/exspy/release_info.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+copyright = "Copyright 2016-2023, The exspy developers"
```
Comment (jlaehne) [inline exspy/exspy/release_info.py]:
```suggestion
copyright = "Copyright 2016-2023, The exSpy developers"
```

Comment (jlaehne) [reply exspy/exspy/release_info.py]: Why from 2016? It is based on eelsLab, which is much older.

File: exspy/exspy/signals/__init__.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+
+""" Signals to be operated on. The basic unit of data"
+    EELSSpectrum
+        For electron energy-loss data with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
Comment (jlaehne) [inline exspy/exspy/signals/__init__.py]:
```suggestion
        For electron energy-loss data with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

File: exspy/exspy/signals/__init__.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    EDSTEMSpectrum
+        For electron energy-dispersive X-rays data acquired in a transmission
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
Comment (jlaehne) [inline exspy/exspy/signals/__init__.py]:
```suggestion
        For electron energy-dispersive X-ray data acquired in a transmission
        electron microscope with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

File: exspy/exspy/signals/__init__.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    EDSSEMSpectrum
+        For electron energy-dispersive X-rays data acquired in a scanning
+        electron microscopy with signal_dimension equal 1, i.e.
+        spectral data of n-dimensions. The signal is binned by default.
```
Comment (jlaehne) [inline exspy/exspy/signals/__init__.py]:
```suggestion
        For electron energy-dispersive X-ray data acquired in a scanning
        electron microscop with signal_dimension equal one, i.e.
        spectral data of `n` navigation dimensions. The signal is binned by default.
```

File: exspy/exspy/signals/__init__.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    DielectricFunction
+        For dielectric function data with signal_dimension equal 1. The signal
+        is unbinned by default.
```
Comment (jlaehne) [inline exspy/exspy/signals/__init__.py]:
```suggestion
        For dielectric function data with signal_dimension equal one. The signal
        is unbinned by default.
```

**Comment by jlaehne** [comment]
References in HyperSpy's `CHANGES.rst` to methods that are now in exSpy will have to use intersphinx.

**Comment by ericpre** [comment]
Yes, for the changelog, it will need updating once the exspy's documentation is live. There are some follow up tweaks needed for the exspy's documentation, but it is not convenient to work on this in this PR. What's matter here is to transfer the relevant bits of code between hyperspy and exspy and as far as I can tell, this should be all done!

**Comment by ericpre** [comment]
Actually some tests could be moved from hyperspy to exspy.

**Comment by francisco-dlp** [comment]
Awesome job!

I think that there are a few more things to do here:

- [ ] Move the convolution feature to exSpy: this sort of convolution only makes sense for CL EELS
- [x] Move the following photoelectron spectroscopy compoents to exSpy:
    - [x] `pes_core_line_shape`
    - [x] `pes_see`
    - [x] `pes_voigt`
 - [ ] Audit the code to identify any API changes worth doing at this point such as privatizing attributes (e.g., `channel_switches`) and methods.

I can work on this next week, as long as somebody is not already working on any of these points.

**Comment by ericpre** [comment]
> * [ ]  Move the convolution feature to exSpy: this sort of convolution only makes sense for CL EELS

Could it be useful in other scenario, for example, to convolve with a point source? The convolution feature itself is generic and not signal specific - even, its know usage is for `EELSModel`.

> * [ ]  Audit the code to identify any API changes worth doing at this point such as privatizing attributes (e.g., `channel_switches`) and methods.

Do you have something else in mind?

**Comment by francisco-dlp** [comment]
Thans for moving the PES components.

While convolving the model could be useful beyond EELS, the way in which the feature is implemented is EELS specific. In particular, the model is evaluated on an axis customized for EELS core-loss (extended to the left to avoid border artifacts). One possibility could be moving only the method that generates the convolution axis to exSpy or (better) generalizing that method to make it useful beyond EELS, and calling it with the right parameters in exSpy.

**Comment by ericpre** [comment]
> One possibility could be moving only the method that generates the convolution axis to exSpy or (better) generalizing that method to make it useful for beyond EELS, and calling it with the right parameters in exSpy.

Yes, it makes sense, and should fairly simple to update/move `set_convolution_axis`? https://github.com/hyperspy/hyperspy/blob/a3a038194d9168b2535c9f1b4003a6b953b93c5a/hyperspy/models/model1d.py#L332-L343C50

I did the `channel_switches` bits, is there anything else that will be needed in this PR?

**Comment by jlaehne** [comment]
> While convolving the model could be useful beyond EELS, but the way in which the feature is implemented is EELS specific.

Indeed, in LumiSpy, we are interested in creating components for luminescence transients that involve a convolution. So it might be useful if that functionality could generalized.

**Comment by francisco-dlp** [comment]
@jlaehne, can you give more details on the convolution feature that you need in LumiSpy?

The current feature convolves a `Signal` with the model of the `Signal` that we are fitting. Both signals must have the same navigation dimensions. Furthermore, the model is evaluated on an axis that is extended to the left, then convolved and cropped to the original axis. This is all very core-loss EELS specific, so I still think that it might be better to move the feature to exSpy.

For general usage, I could imagine convolving the whole model with a component (typically a peaky one, such as a Gaussian). The parameters in this case could be free (something not possible with the current implementation). However, nobody has yet requested such  a feature...

**Comment by jlaehne** [comment]
> can you give more details on the convolution feature that you need in LumiSpy?

In LumiSpy, we would want to convolute a system response function with a decay model.

The system response function could be
- a Gaussian or
- a hyperbolic secant component,
- but possibly also a measured curve (so a `signal` as in the EELS case). The decay model could 

The decay model could be e.g.
- one or multiple exponentials
- a stretched exponential
- ...

Usually, the rise time of the system response is included in the experimental data, so we probably would not have to extend the model as in EELS, but I might have a closer look at your implementation.

**Comment by francisco-dlp** [comment]
Thanks for the details @jlaehne. I have some questions:

- Is the decay model part of a larger model?
- When the system response is a signal, is it different at each pixel?

**Comment by jlaehne** [comment]
>     * Is the decay model part of a larger model?
As indicated above, the model can have 1-3 components for the decay (and that depends on the sample, but would usually be the same across a map) and this is convolved with the system response. The parameters of these components change of course from pixel to pixel. So far, I do not see any overarching larger model.

>     * When the system response is a signal, is it different at each pixel?
No, the system response function should be the same for every pixel in the navigation dimension.

**Comment by francisco-dlp** [comment]
In that case, I think that the convolution feature implemented for EELS won't be of much use for your application.

**Comment by ericpre** [comment]
@francisco-dlp, do you know if you would have today or tomorrow to do the convolution bits? If not, I will try to do that later today in the interest of finishing this PR.
As I understand, the approach currently used for setting the convolution axis is valid but suboptimal (1. larger axis due to extrapolation makes slower fit, 2. this is not implementation for linear fitting without iterating over the navigation space) in more simple convolution case, where no extrapolation is necessary. If my understand if correct, we simply need to move the `set_convolution_axis` to `EELSModel` and make a more simple version in `Model1D`, using the `Model1D.axis.axis` without extrapolation?
The improvement with linear fitting can be done in a separate PR, as this is unrelated to the exSpy split.

**Comment by francisco-dlp** [comment]
I can start working on it tomorrow.

I still think that, as it is, the feature is too EELS-specific. Therefore, I propose to move it all to exSpy. If somebody needs convolving the whole model in the future, chances are that we'll need to take a different approach anyway since most usage cases involve convolving with a single signal that could possible be a component with free parameters. Does it make sense?

**Comment by ericpre** [comment]
Aside the axis/model extrapolation, what else is left that is EELS specific? With moving it all to exSpy, do you mean removing the convolution functionality from `Model1D`? We just need to add supports for single signal, so that it can be more straightforward to use with more simple case than EELS. Then it will be easy to implement the convolution for linear fit without iteration over the navigation.

**Comment by francisco-dlp** [comment]
Yes, I mean removing the convolution functionality from `Model1D`. This is because, if we simply more `set_convolution_axis`, the rest of the code is not usable as is. Therefore, I propose to move all to exSpy. This has the advantage that we can rename `convolve_signal` back to `low_loss` to enhance backward compatibility. If HyperSpy ever needs a convolution function based on the existing one we can always move it back.

**Comment by CSSFrancis** [comment]
@ericpre and @francisco-dlp what is left that needs to be done here?  I'm for moving the convolution stuff to `exspy` if that is the thing holding things up.

**Comment by francisco-dlp** [comment]
@CSSFrancis, I was planning to do it myself. I have already started, but I won't be able to come back to it until Thursday afternoon. If you have already started yourself and you feel like it, go on, but do let me know.

**Comment by CSSFrancis** [comment]
> @CSSFrancis, I was planning to do it myself. I have already started, but I won't be able to come back to it until Thursday afternoon. If you have already started yourself and you feel like it, go on, but do let me know.

I haven't started anything yet :) It would probably would take me a little bit to familiarize myself with the code, just let me know if you need me to look at anything or review any additions.

**Comment by ericpre** [comment]
@francisco-dlp, do you known when you will be able to move the convolution to exspy? Looking at what needs to be done, it seems not straightforward as it is used in the calculation of the jacobian, and other places. If it can't done in a reasonable timeline, I would suggest to leave it for later, because this is holding the progress toward releasing hyperspy 2.0 - what is left should be some tidying up to be done after exspy is split.

**Comment by francisco-dlp** [comment]
Yes, it is not straightforward. I am finishing it. I will submit a PR to this branch today.

**Comment by francisco-dlp** [comment]
Here is the PR: https://github.com/CSSFrancis/hyperspy/pull/2

Moving the code required some small structural changes that should make it easier to implement convolution in other `BaseModel` subclasses.

Some convolution code remains in HyperSpy:

- The `Component.convolved` attribute
- The `Model.convolved` attribute for back-compatibility
- Linear fit code for convolution in `Expression` (there is no way to move it elsewhere)
- A few harmless lines of code in SAMFIRE check the `Model.convolved` attribute among others. Since that part of the code is not covered by tests, and SAMFIRE requires some care anyway, I prefer no to attempt to move it at this stage.

**Comment by CSSFrancis** [comment]
@francisco-dlp I (think) I fixed the failing tests.  One was pointing to a file which didn't exist.  Loading some of the older Models was also failing if you didn't have `exspy` installed and wasn't pointing towards installing `exspy` so I fixed that.

**Comment by CSSFrancis** [comment]
@ericpre The failure is related to the pooch/zendo bug correct?

https://github.com/fatiando/pooch/issues/371 We should be able to run the CI again once pooch 1.8.0 hits conda in an hour or two.

**Comment by CSSFrancis** [comment]
Thoughts on merging/splitting and then figuring stuff like the documentation out separately?

**Review by ericpre — APPROVED** [review]
Azure pipelines CI is still failing because pooch 1.8 is still not on conda-forge, it should be a couple of hours.

@francisco-dlp, do you want to do another review?

**Comment by CSSFrancis** [comment]
@ericpre  @francisco-dlp Do we want to merge this and then we can try to clean some stuff in later PR's if needed.  I think this PR is mostly good (at least 3 people have looked at the changes and the CI passes.)

**Comment by ericpre** [comment]
I would suggest to leave it up open for comments to tomorrow, and if this is none, then we merge?

File: hyperspy/model.py [hunk 11]
```diff
@@ -126,11 +126,18 @@ def reconstruct_component(comp_dictionary, **init_args):
...
-            'failed because the component is provided by the ' +
-            f'{comp_dictionary["package"]} Python package, but ' +
-            f'{comp_dictionary["package"]} is not installed.')
+        try:
+            raise ImportError(
+                f'Loading the {comp_dictionary["class"]} component ' +
+                'failed because the component is provided by the ' +
+                f'{comp_dictionary["package"]} Python package, but ' +
+                f'{comp_dictionary["package"]} is not installed.')
+        except KeyError:
```
Comment (francisco-dlp) [inline hyperspy/model.py]: Would it be possible to make this more precise e.g. by checking if the component's package is HyperSpy and `comp_dictionary["class"]` is one of:

```python
EXSPY_HSPY_COMPONENTS = ("EELSArctan", "EELSCLEdge", "DoublePowerLaw", "Vignetting", "PESCoreLineShape", "SEE", "PESVoigt", "VolumePlasmonDrude")
```

?

**Comment by francisco-dlp** [comment]
I have sent a PR (https://github.com/CSSFrancis/hyperspy/pull/3) to fix the path of the file that @CSSFrancis though that we have moved to exSpy.

Other than this and the small comment above, this LGTM!

**Comment by CSSFrancis** [comment]
> I have sent a PR ([CSSFrancis#3](https://github.com/CSSFrancis/hyperspy/pull/3)) to fix the path of the file that @CSSFrancis though that we have moved to exSpy.
> 
> Other than this and the small comment above, this LGTM!

Done!

**Comment by CSSFrancis** [comment]
LGTM?

**Comment by CSSFrancis** [comment]
@ericpre Do you want to merge this and do the git filter?  I can give it another shot otherwise but I'm not sure how great I did last time 😅

**Comment by ericpre** [comment]
I would like to check briefly the last commit and #3238 also needs to be merge before splitting.

I am happy to do it in the coming days.

**Comment by ericpre** [comment]
For reference, this is the command that I have used to split to https://github.com/hyperspy/exspy:
```
git filter-repo --path exspy --path spectrospy --path hyperspy/tests/utils --path hyperspy/tests/misc --path hyperspy/tests/models --path hyperspy/tests/components --path hyperspy/tests/drawing/plot_model --path hyperspy/tests/drawing/plot_spectra_markers --path hyperspy/tests/drawing/data --path hyperspy/tests/drawing/test_plot_spectra_markers.py --path hyperspy/tests/drawing/test_plot_model.py --path hyperspy/tests/signals --path hyperspy/signal_tools.py --path hyperspy/docstrings/model.py --path hyperspy/misc/elements.py --path hyperspy/misc/material.py --path hyperspy/misc/eds  --path hyperspy/misc/eels --path examples/model_fitting --path hyperspy/_signals --path hyperspy/signals --path hyperspy/_components --path hyperspy/components --path eelslab --path lib --path src --path-rename exspy/: --force
```
It is quite long to cherry pick the relevant folders from the history and remove the files that are not relevant (and sometimes large, e.g. io test file, etc.).

**Pull request merged**